Detectors/TOF/base/src/Geo.cxx

```diff
@@ -19,3 +20,478 @@ constexpr Float_t Geo::HEIGHTS[NPLATES][NMAXNSTRIP];
 constexpr Float_t Geo::DISTANCES[NPLATES][NMAXNSTRIP];
 constexpr Bool_t Geo::FEAWITHMASKS[18];
 constexpr Float_t Geo::ROOF2PARAMETERS[3];
+
+void Geo::GetDetID(Float_t* pos, Int_t* det)
+{
+  //
+  // Returns Detector Indices (iSect,iPlate,iStrip,iPadX,iPadZ)
+  // space point coor (x,y,z) (cm)
+
+  det[0] = GetSector(pos);
+  det[1] = GetPlate(pos);
+  det[2] = GetStrip(pos);
+  det[3] = GetPadZ(pos);
+  det[4] = GetPadX(pos);
+}
+
+Int_t Geo::GetSector(const Float_t* pos)
+{
+  //
+  // Returns the Sector index
+  //
+
+  Int_t iSect = -1;
+
+  Float_t x = pos[0];
+  Float_t y = pos[1];
+  Float_t z = pos[2];
+
+  Float_t rho = TMath::Sqrt(x * x + y * y);
+
+  if (!((z >= -ZLENA * 0.5 && z <= ZLENA * 0.5) && (rho >= (RMIN) && rho <= (RMAX)))) {
+    // AliError("Detector Index could not be determined");
+    return iSect;
+  }
+
+  Float_t phi = TMath::Pi() + TMath::ATan2(-y, -x);
+
+  iSect = (Int_t)(phi * TMath::RadToDeg() / PHISEC);
+
+  return iSect;
+}
+
+Int_t Geo::GetPlate(const Float_t* pos)
+{
+  //
+  // Returns the Plate index
+  //
+
+  Int_t iPlate = -1;
+
+  Float_t posLocal[3];
+  for (Int_t ii = 0; ii < 3; ii++)
+    posLocal[ii] = pos[ii];
+
+  Int_t isector = GetSector(posLocal);
+  if (isector == -1) {
+    // AliError("Detector Index could not be determined");
+    return iPlate;
+  }
+
+  // ALICE reference frame -> B071/B074/B075 = BTO1/2/3 reference frame
+  Double_t angles[6] = { 90., 90. + (isector + 0.5) * PHISEC, 0., 0., 90., (isector + 0.5) * PHISEC };
```

Comment:
This sequence of Rotations and Translation is done with lot of calculations (partucilarly building the matrices with plenty of trig. fun. calls), while the set of parameters for rotations/translations is limited. I would create these matrices in advance (as far as I see, array of matrices, 1 per sector) and use transformation routines of the matrices. I am going to commit soon the 3D transformation class (functionality similar to TGeoHMatrix) which can operate both on double and float 3d point classes.